# Illegal Postings

In [1]:
import pandas as pd 
import numpy as np

In [2]:
df = pd.read_csv('../../data/raw/Illegal_Postings.csv')

In [3]:
df.head()

CaseID                  Opened Closed                 Updated Status  \
0  11771777  12/04/2019 07:56:00 PM    NaN  12/04/2019 08:30:10 PM   Open   
1  11767854  12/04/2019 07:56:00 AM    NaN  12/04/2019 08:20:36 AM   Open   
2  11771943  12/04/2019 08:53:00 PM    NaN  12/04/2019 09:40:05 PM   Open   
3  11771937  12/04/2019 08:51:00 PM    NaN  12/04/2019 09:40:04 PM   Open   
4  11771923  12/04/2019 08:44:00 PM    NaN  12/04/2019 09:25:05 PM   Open   

  Status Notes Responsible Agency          Category  \
0     accepted      DPW Ops Queue  Illegal Postings   
1     accepted      DPW Ops Queue  Illegal Postings   
2     accepted      DPW Ops Queue  Illegal Postings   
3     accepted      DPW Ops Queue  Illegal Postings   
4     accepted      DPW Ops Queue  Illegal Postings   

                                 Request Type          Request Details  \
0  Illegal Postings - Posted_on_Traffic_Light  Posted on Traffic Light   
1       Illegal Postings - Affixed_Improperly       Affixed Improperly   
2  Illegal Postings - Posted_on_Traffic_Light  Posted on Traffic Light   
3  Illegal Postings - Posted_on_Traffic_Light  Posted on Traffic Light   
4       Illegal Postings - Affixed_Improperly       Affixed Improperly   

                                   Address    Street  Supervisor District  \
0   2100 PINE ST, SAN FRANCISCO, CA, 94115   PINE ST                  2.0   
1   1283 9TH AVE, SAN FRANCISCO, CA, 94122   9TH AVE                  5.0   
2   990 JONES ST, SAN FRANCISCO, CA, 94109  JONES ST                  3.0   
3    901 PINE ST, SAN FRANCISCO, CA, 94108   PINE ST                  3.0   
4  1202 JONES ST, SAN FRANCISCO, CA, 94109  JONES ST                  3.0   

      Neighborhood Police District   Latitude   Longitude  \
0  Pacific Heights        NORTHERN  37.788520 -122.430566   
1     Inner Sunset         TARAVAL  37.764323 -122.466583   
2   Lower Nob Hill         CENTRAL  37.790344 -122.413654   
3   Lower Nob Hill         CENTRAL  37.790764 -122.410833   
4         Nob Hill         CENTRAL  37.792301 -122.414105   

                        Point Source Media URL  
0  (37.7885202, -122.4305658)    Web       NaN  
1  (37.7643228, -122.4665826)    Web       NaN  
2  (37.7903441, -122.4136543)    Web       NaN  
3  (37.7907644, -122.4108334)    Web       NaN  
4   (37.792301, -122.4141046)    Web       NaN

In [4]:
df.dtypes

CaseID                   int64
Opened                  object
Closed                  object
Updated                 object
Status                  object
Status Notes            object
Responsible Agency      object
Category                object
Request Type            object
Request Details         object
Address                 object
Street                  object
Supervisor District    float64
Neighborhood            object
Police District         object
Latitude               float64
Longitude              float64
Point                   object
Source                  object
Media URL               object
dtype: object

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59895 entries, 0 to 59894
Data columns (total 20 columns):
CaseID                 59895 non-null int64
Opened                 59895 non-null object
Closed                 58693 non-null object
Updated                59895 non-null object
Status                 59895 non-null object
Status Notes           40257 non-null object
Responsible Agency     59895 non-null object
Category               59895 non-null object
Request Type           59895 non-null object
Request Details        59894 non-null object
Address                59895 non-null object
Street                 59756 non-null object
Supervisor District    59756 non-null float64
Neighborhood           59694 non-null object
Police District        59676 non-null object
Latitude               59895 non-null float64
Longitude              59895 non-null float64
Point                  59895 non-null object
Source                 59895 non-null object
Media URL              29183 non-n

## Rename columns

In [6]:
df.rename(columns={'Latitude': 'Y',
                  'Longitude': 'X'}, inplace=True)
df.head()

CaseID                  Opened Closed                 Updated Status  \
0  11771777  12/04/2019 07:56:00 PM    NaN  12/04/2019 08:30:10 PM   Open   
1  11767854  12/04/2019 07:56:00 AM    NaN  12/04/2019 08:20:36 AM   Open   
2  11771943  12/04/2019 08:53:00 PM    NaN  12/04/2019 09:40:05 PM   Open   
3  11771937  12/04/2019 08:51:00 PM    NaN  12/04/2019 09:40:04 PM   Open   
4  11771923  12/04/2019 08:44:00 PM    NaN  12/04/2019 09:25:05 PM   Open   

  Status Notes Responsible Agency          Category  \
0     accepted      DPW Ops Queue  Illegal Postings   
1     accepted      DPW Ops Queue  Illegal Postings   
2     accepted      DPW Ops Queue  Illegal Postings   
3     accepted      DPW Ops Queue  Illegal Postings   
4     accepted      DPW Ops Queue  Illegal Postings   

                                 Request Type          Request Details  \
0  Illegal Postings - Posted_on_Traffic_Light  Posted on Traffic Light   
1       Illegal Postings - Affixed_Improperly       Affixed Improperly   
2  Illegal Postings - Posted_on_Traffic_Light  Posted on Traffic Light   
3  Illegal Postings - Posted_on_Traffic_Light  Posted on Traffic Light   
4       Illegal Postings - Affixed_Improperly       Affixed Improperly   

                                   Address    Street  Supervisor District  \
0   2100 PINE ST, SAN FRANCISCO, CA, 94115   PINE ST                  2.0   
1   1283 9TH AVE, SAN FRANCISCO, CA, 94122   9TH AVE                  5.0   
2   990 JONES ST, SAN FRANCISCO, CA, 94109  JONES ST                  3.0   
3    901 PINE ST, SAN FRANCISCO, CA, 94108   PINE ST                  3.0   
4  1202 JONES ST, SAN FRANCISCO, CA, 94109  JONES ST                  3.0   

      Neighborhood Police District          Y           X  \
0  Pacific Heights        NORTHERN  37.788520 -122.430566   
1     Inner Sunset         TARAVAL  37.764323 -122.466583   
2   Lower Nob Hill         CENTRAL  37.790344 -122.413654   
3   Lower Nob Hill         CENTRAL  37.790764 -122.410833   
4         Nob Hill         CENTRAL  37.792301 -122.414105   

                        Point Source Media URL  
0  (37.7885202, -122.4305658)    Web       NaN  
1  (37.7643228, -122.4665826)    Web       NaN  
2  (37.7903441, -122.4136543)    Web       NaN  
3  (37.7907644, -122.4108334)    Web       NaN  
4   (37.792301, -122.4141046)    Web       NaN

## Empty values

In [7]:
for col in df:
    print(col, df[col].isnull().sum() )

CaseID 0
Opened 0
Closed 1202
Updated 0
Status 0
Status Notes 19638
Responsible Agency 0
Category 0
Request Type 0
Request Details 1
Address 0
Street 139
Supervisor District 139
Neighborhood 201
Police District 219
Y 0
X 0
Point 0
Source 0
Media URL 30712


Closed nan values are fine, because the case is still open and we will only look at the open date. Status notes and request type are also not useful. Basically we need to take care of nan values for Location, Supervisor District, Neighborhood and Police District

In [8]:
df[df['Supervisor District'].isnull()]

CaseID                  Opened                  Closed  \
850    11679619  11/12/2019 02:23:00 PM  11/16/2019 01:02:56 PM   
870    11679847  11/12/2019 02:54:00 PM  11/15/2019 10:31:50 AM   
1104   11669360  11/10/2019 09:54:00 AM  11/10/2019 01:03:33 PM   
2179   11533309  10/10/2019 02:28:00 PM  10/20/2019 10:00:06 PM   
2722   11533395  10/10/2019 02:42:00 PM  10/10/2019 10:10:31 PM   
2880   11513770  10/06/2019 06:06:00 PM  10/08/2019 05:05:00 PM   
3387    4024617  09/18/2014 10:47:14 AM  09/18/2014 11:22:46 AM   
3420    5920742  05/29/2016 05:05:26 PM  06/02/2016 12:00:00 PM   
3463    4183950  11/06/2014 03:41:48 PM  11/12/2014 05:00:00 PM   
3508    1101239  04/09/2012 04:33:22 AM  04/25/2012 12:12:00 PM   
3526    3872526  07/30/2014 09:31:38 AM  07/30/2014 07:58:38 AM   
3530    3846024  07/21/2014 12:08:16 PM  07/21/2014 12:31:28 PM   
3632    2673414  08/03/2013 09:52:19 AM  08/03/2013 11:02:41 AM   
3633    2445992  06/03/2013 04:54:47 PM                     NaN   
3634    2427344  05/30/2013 07:33:56 AM  05/30/2013 09:20:23 AM   
3653     881132  05/13/2011 12:01:45 PM  05/16/2011 06:06:03 AM   
3768     339165  12/26/2008 02:12:20 PM  12/27/2008 06:07:08 AM   
3769     684323  06/29/2010 02:44:38 PM  07/21/2010 02:17:01 PM   
3783     308535  11/04/2008 06:59:15 AM  11/05/2008 01:58:59 PM   
3832     430390  05/13/2009 06:35:46 AM  05/30/2009 04:10:52 PM   
3874     376541  02/22/2009 05:25:20 PM  02/23/2009 02:10:02 PM   
3924     243532  07/07/2008 11:10:57 AM  07/07/2008 04:35:39 PM   
3968     329459  12/10/2008 10:13:16 AM  12/12/2008 06:07:20 AM   
4002     317220  11/19/2008 05:43:39 PM  11/21/2008 06:07:02 AM   
4052     247311  07/14/2008 11:53:37 AM  02/19/2009 01:09:57 PM   
4133     387014  03/10/2009 06:17:23 PM  03/12/2009 03:10:11 PM   
4207     380145  02/28/2009 10:14:02 AM  03/26/2009 03:10:15 PM   
4275     308917  11/04/2008 04:07:11 PM  11/04/2008 06:06:52 PM   
4287     368909  02/09/2009 11:28:43 AM  02/09/2009 01:09:53 PM   
4304     485149  08/11/2009 07:50:02 AM  08/19/2009 11:11:14 AM   
...         ...                     ...                     ...   
22046   4369839  01/07/2015 09:25:36 AM  01/18/2015 02:20:00 AM   
22452   4190031  11/08/2014 03:43:05 PM  05/12/2015 04:24:50 PM   
22921   4173223  11/03/2014 03:35:21 PM  01/14/2015 10:20:00 AM   
23626   3949761  08/25/2014 12:01:21 PM  08/26/2014 03:58:20 AM   
24248   4026142  09/18/2014 04:30:37 PM  09/19/2014 04:41:42 AM   
24352   3906959  08/11/2014 11:15:57 AM  08/11/2014 11:59:58 AM   
26798   4689418  04/23/2015 09:38:35 AM  04/24/2015 12:00:00 PM   
26863   4589370  03/19/2015 12:48:22 PM  03/20/2015 07:41:55 AM   
29184   5137532  09/28/2015 12:05:49 AM  02/25/2016 12:00:00 PM   
30585   5168411  10/07/2015 03:27:34 PM  10/09/2015 12:00:00 PM   
30614   4990562  08/12/2015 12:42:31 AM  08/19/2015 12:00:00 PM   
31102   4902719  07/12/2015 06:31:11 PM  07/14/2015 12:00:00 PM   
33659   5462633  01/13/2016 04:06:47 PM  01/27/2016 12:00:00 PM   
33760   5297546  11/18/2015 03:02:33 PM  11/20/2015 12:00:00 PM   
33849   5456894  01/12/2016 10:11:27 AM  01/13/2016 12:00:00 PM   
34161   5583469  02/19/2016 10:23:22 AM  02/29/2016 12:00:00 PM   
34851   5466163  01/14/2016 03:43:06 PM  01/15/2016 12:00:00 PM   
35224   5600351  02/24/2016 03:11:09 PM  11/10/2016 01:24:41 AM   
35399   5761321  04/13/2016 10:25:31 PM  04/21/2016 12:00:00 PM   
35498   5676707  03/17/2016 09:38:31 PM  03/25/2016 12:00:00 PM   
36184   5676741  03/17/2016 10:01:50 PM  03/26/2016 12:00:00 PM   
37068   5690494  03/22/2016 05:03:49 PM  12/15/2016 11:46:28 AM   
39442   6129336  07/27/2016 11:07:42 PM  02/12/2017 05:11:24 AM   
40826   6144817  08/01/2016 02:23:19 PM  08/05/2016 01:12:22 AM   
48996   8869061  04/13/2018 03:48:00 PM  04/17/2018 12:25:20 AM   
50843   9309397  07/24/2018 11:28:00 AM  07/24/2018 11:29:00 AM   
56448  11205381  07/31/2019 10:24:00 AM  07/31/2019 12:32:40 PM   
57749  11393224  09/09/2019 09:09:00 PM  09/10/2019 12:0

We see a problem. The rows with no district are the ones with location (0,0). Since we are going to work based on locations. Drop all the locations with (0,0)

In [9]:
df = df[df.Point != '(0.0, 0.0)']
df[df['Supervisor District'].isnull()]

CaseID                  Opened                  Closed  \
850    11679619  11/12/2019 02:23:00 PM  11/16/2019 01:02:56 PM   
870    11679847  11/12/2019 02:54:00 PM  11/15/2019 10:31:50 AM   
1104   11669360  11/10/2019 09:54:00 AM  11/10/2019 01:03:33 PM   
2179   11533309  10/10/2019 02:28:00 PM  10/20/2019 10:00:06 PM   
2722   11533395  10/10/2019 02:42:00 PM  10/10/2019 10:10:31 PM   
2880   11513770  10/06/2019 06:06:00 PM  10/08/2019 05:05:00 PM   
18868   3673285  05/24/2014 10:02:27 AM  05/24/2014 11:48:25 AM   
24248   4026142  09/18/2014 04:30:37 PM  09/19/2014 04:41:42 AM   
57749  11393224  09/09/2019 09:09:00 PM  09/10/2019 12:07:44 PM   
58165  11393177  09/09/2019 08:55:00 PM  09/09/2019 10:10:20 PM   
58242  11393183  09/09/2019 08:56:00 PM  09/10/2019 08:25:42 AM   

                      Updated  Status  \
850    11/16/2019 01:02:56 PM  Closed   
870    11/15/2019 10:31:50 AM  Closed   
1104   11/10/2019 01:03:33 PM  Closed   
2179   10/20/2019 10:00:06 PM  Closed   
2722   10/10/2019 10:10:31 PM  Closed   
2880   10/08/2019 05:05:00 PM  Closed   
18868  05/24/2014 11:48:25 AM  Closed   
24248  09/19/2014 04:41:42 AM  Closed   
57749  09/10/2019 12:07:44 PM  Closed   
58165  09/09/2019 10:10:20 PM  Closed   
58242  09/10/2019 08:25:42 AM  Closed   

                                            Status Notes Responsible Agency  \
850                                                  NaN      DPW Ops Queue   
870      SES Graffiti Crew  - Scrape - Misc - Other Work      DPW Ops Queue   
1104                                                 NaN      DPW Ops Queue   
2179   Closed: No response from PG & E- GRAFFITI in 1...      DPW Ops Queue   
2722   Request is a duplicate and has been previously...      DPW Ops Queue   
2880                                                 NaN      DPW Ops Queue   
18868                               resolved: 1098 - cmp      DPW Ops Queue   
24248                                     resolved: 1098      DPW Ops Queue   
57749                                                NaN      DPW Ops Queue   
58165  Request is a duplicate and has been previously...      DPW Ops Queue   
58242                                                NaN      DPW Ops Queue   

               Category                                Request Type  \
850    Illegal Postings       Illegal Postings - Affixed_Improperly   
870    Illegal Postings       Illegal Postings - Affixed_Improperly   
1104   Illegal Postings       Illegal Postings - Affixed_Improperly   
2179   Illegal Postings       Illegal Postings - Affixed_Improperly   
2722   Illegal Postings       Illegal Postings - Affixed_Improperly   
2880   Illegal Postings       Illegal Postings - Affixed_Improperly   
18868  Illegal Postings       Illegal Postings - Affixed_Improperly   
24248  Illegal Postings        Illegal Postings - Multiple_Postings   
57749  Illegal Postings       Illegal Postings - Affixed_Improperly   
58165  Illegal Postings  Illegal Postings - Posted_on_Traffic_Light   
58242  Illegal Postings  Illegal Postings - Posted_on_Traffic_Light   

                     Request Details  \
850               Affixed Improperly   
870               Affixed Improperly   
1104              Affixed Improperly   
2179              Affixed Improperly   
2722              Affixed Improperly   
2880              Affixed Improperly   
18868    Affixed_Improperly on Other   
24248  Multiple_Postings on Sidewalk   
57749             Affixed Improperly   
58165        Posted on Traffic Light   
58242        Posted on Traffic Light   

                                            Address Street  \
850        2516 BALBOA ST, SAN FRANCISCO, CA, 94121    NaN   
870        2516 BALBOA ST, SAN FRANCISCO, CA, 94121    NaN   
1104   1445 SAN BRUNO AVE, SAN FRANCISCO, CA, 94110    NaN   
2179       106 SHIELDS ST, SAN FRANCISCO, CA, 94132    NaN   
2722       499 SHIELDS ST, SAN FRANCISCO, CA, 94132    NaN   
2880       1395 HAIGHT ST, SAN FRANCISCO, CA, 94117    N

We got rid from a lot of missing location rows, but we are still missing some supervisor districts. We can get the correct district using the location later with our class.

In [10]:
print('Nan on Neighborhood', df[df['Neighborhood'].isnull()].shape[0])
print('Nan on Police District', df[df['Police District'].isnull()].shape[0])

Nan on Neighborhood 63
Nan on Police District 81


In [11]:
df[df['Neighborhood'].isnull()]

CaseID                  Opened                  Closed  \
850    11679619  11/12/2019 02:23:00 PM  11/16/2019 01:02:56 PM   
870    11679847  11/12/2019 02:54:00 PM  11/15/2019 10:31:50 AM   
1104   11669360  11/10/2019 09:54:00 AM  11/10/2019 01:03:33 PM   
2179   11533309  10/10/2019 02:28:00 PM  10/20/2019 10:00:06 PM   
2722   11533395  10/10/2019 02:42:00 PM  10/10/2019 10:10:31 PM   
2880   11513770  10/06/2019 06:06:00 PM  10/08/2019 05:05:00 PM   
3655    5144672  09/29/2015 06:44:27 PM  01/12/2016 12:00:00 PM   
18868   3673285  05/24/2014 10:02:27 AM  05/24/2014 11:48:25 AM   
21296   3575162  04/22/2014 07:54:51 AM  04/22/2014 08:10:34 AM   
24248   4026142  09/18/2014 04:30:37 PM  09/19/2014 04:41:42 AM   
25310   4603896  03/24/2015 05:21:05 PM  03/25/2015 08:19:35 AM   
27466   4604015  03/24/2015 05:47:38 PM  03/25/2015 08:19:35 AM   
27547   4604115  03/24/2015 06:23:00 PM  03/25/2015 08:19:07 AM   
28181   4648493  04/08/2015 05:31:41 PM  04/09/2015 01:30:00 AM   
28428   4468315  02/06/2015 10:33:40 PM  02/12/2015 12:00:00 PM   
28634   5021788  08/21/2015 04:13:48 PM  08/21/2015 05:05:56 PM   
30191   5144567  09/29/2015 06:06:42 PM  02/25/2016 12:00:00 PM   
30419   5172039  10/08/2015 02:25:24 PM  10/11/2015 02:00:13 AM   
30831   5144535  09/29/2015 05:58:13 PM  01/12/2016 12:00:00 PM   
31085   5021796  08/21/2015 04:14:49 PM  08/21/2015 05:06:11 PM   
31205   4826649  06/14/2015 04:46:14 PM  07/21/2015 12:00:00 PM   
33047   5021802  08/21/2015 04:16:09 PM  08/21/2015 05:06:32 PM   
33216   5294750  11/17/2015 07:54:10 PM  01/12/2016 12:00:00 PM   
33509   5436364  01/05/2016 09:47:29 PM  01/13/2016 12:00:00 PM   
33518   5486647  01/20/2016 09:45:28 PM  05/09/2016 04:36:19 PM   
33885   5493458  01/22/2016 04:01:18 PM  12/15/2016 10:52:21 AM   
34965   5486029  01/20/2016 05:10:36 PM  05/09/2016 04:36:28 PM   
35095   5267844  11/09/2015 10:46:39 AM  12/08/2015 12:00:00 PM   
35136   5294772  11/17/2015 08:03:53 PM  01/12/2016 12:00:00 PM   
35192   5294752  11/17/2015 07:54:53 PM  01/12/2016 12:00:00 PM   
...         ...                     ...                     ...   
36611   5897857  05/23/2016 11:20:40 AM  06/14/2016 02:17:12 AM   
37143   5967555  06/10/2016 04:09:38 PM                     NaN   
37186   5720981  04/01/2016 12:10:13 PM  04/08/2016 01:34:40 AM   
37426   5877702  05/17/2016 02:51:25 PM  06/14/2016 02:13:47 AM   
37742   5867629  05/14/2016 06:58:12 PM  05/17/2016 02:46:31 AM   
37987   5897861  05/23/2016 11:21:16 AM  05/31/2016 08:17:03 AM   
38090   5993807  06/18/2016 02:23:32 PM  07/06/2016 09:28:30 AM   
38908   6294024  09/09/2016 03:18:50 PM  10/18/2016 02:01:13 AM   
38950   6294155  09/09/2016 03:42:49 PM  01/08/2017 06:25:55 AM   
39205   6294005  09/09/2016 03:16:14 PM  10/18/2016 01:52:34 AM   
39734   6157457  08/04/2016 12:01:17 PM  08/17/2016 10:10:49 AM   
39901   6293982  09/09/2016 03:10:01 PM  10/18/2016 02:01:19 AM   
39976   6294120  09/09/2016 03:35:18 PM  10/18/2016 02:00:39 AM   
40281   6293983  09/09/2016 03:10:48 PM  10/18/2016 01:52:43 AM   
40932   6294045  09/09/2016 03:21:48 PM  10/18/2016 01:51:48 AM   
42676   6701063  01/06/2017 10:33:36 PM  01/13/2017 02:15:24 PM   
44238   7230466  06/06/2017 12:44:51 PM  06/10/2017 04:25:16 PM   
46242   8227683  11/04/2017 04:28:00 PM  11/05/2017 01:00:00 AM   
46583   8227660  11/04/2017 04:23:00 PM  11/05/2017 01:00:00 AM   
46983   8063474  09/25/2017 03:24:00 PM  09/26/2017 07:32:00 AM   
47068   8310298  11/26/2017 08:28:00 AM  11/28/2017 12:41:00 PM   
47450   8105259  10/05/2017 12:54:00 PM  10/05/2017 03:05:00 PM   
47878   8375237  12/11/2017 04:45:00 PM  12/29/2017 08:38:00 PM   
48106   8478238  01/08/2018 12:30:00 PM  01/08/2018 01:01:30 PM   
48123   8529005  01/19/2018 06:34:00 PM  01/20/2018 12:00:00 AM   
51510   9500636  09/05/2018 04:33:00 PM  09/12/2018 01:52:00 PM   
54804  10647462  03/26/2019 09:45:00 AM  03/26/2019 10:31:00 AM   
57749  11393224  09/09/2019 09:09:00 PM  09/10/2019 12:0

There are less than 81 rows with nan out of 59895, that is 0.1%. Therefore, we will drop them without risk of losing data

In [12]:
df.dropna(subset=['Neighborhood'], inplace=True)
df.dropna(subset=['Police District'], inplace=True)
df.dropna(subset=['Supervisor District'], inplace=True)

In [13]:
print('Nan on Neighborhood', df[df['Neighborhood'].isnull()].shape[0])
print('Nan on Police District', df[df['Police District'].isnull()].shape[0])
print('Nan on Supervisor District', df[df['Supervisor District'].isnull()].shape[0])

Nan on Neighborhood 0
Nan on Police District 0
Nan on Supervisor District 0


## Datetime objects

In [14]:
df['Opened'] = pd.to_datetime(df['Opened'] )
df['Closed'] = pd.to_datetime(df['Closed'] )
df['Updated'] = pd.to_datetime(df['Updated'] )

In [15]:
df['Year'] = df['Opened'].dt.year
df['Month'] = df['Opened'].dt.month
df['Day'] = df['Opened'].dt.day

In [16]:
df.head()

CaseID              Opened Closed             Updated Status  \
0  11771777 2019-12-04 19:56:00    NaT 2019-12-04 20:30:10   Open   
1  11767854 2019-12-04 07:56:00    NaT 2019-12-04 08:20:36   Open   
2  11771943 2019-12-04 20:53:00    NaT 2019-12-04 21:40:05   Open   
3  11771937 2019-12-04 20:51:00    NaT 2019-12-04 21:40:04   Open   
4  11771923 2019-12-04 20:44:00    NaT 2019-12-04 21:25:05   Open   

  Status Notes Responsible Agency          Category  \
0     accepted      DPW Ops Queue  Illegal Postings   
1     accepted      DPW Ops Queue  Illegal Postings   
2     accepted      DPW Ops Queue  Illegal Postings   
3     accepted      DPW Ops Queue  Illegal Postings   
4     accepted      DPW Ops Queue  Illegal Postings   

                                 Request Type          Request Details  ...  \
0  Illegal Postings - Posted_on_Traffic_Light  Posted on Traffic Light  ...   
1       Illegal Postings - Affixed_Improperly       Affixed Improperly  ...   
2  Illegal Postings - Posted_on_Traffic_Light  Posted on Traffic Light  ...   
3  Illegal Postings - Posted_on_Traffic_Light  Posted on Traffic Light  ...   
4       Illegal Postings - Affixed_Improperly       Affixed Improperly  ...   

      Neighborhood Police District          Y           X  \
0  Pacific Heights        NORTHERN  37.788520 -122.430566   
1     Inner Sunset         TARAVAL  37.764323 -122.466583   
2   Lower Nob Hill         CENTRAL  37.790344 -122.413654   
3   Lower Nob Hill         CENTRAL  37.790764 -122.410833   
4         Nob Hill         CENTRAL  37.792301 -122.414105   

                        Point  Source  Media URL  Year Month Day  
0  (37.7885202, -122.4305658)     Web        NaN  2019    12   4  
1  (37.7643228, -122.4665826)     Web        NaN  2019    12   4  
2  (37.7903441, -122.4136543)     Web        NaN  2019    12   4  
3  (37.7907644, -122.4108334)     Web        NaN  2019    12   4  
4   (37.792301, -122.4141046)     Web        NaN  2019    12   4  

[5 rows x 23 columns]

## Readability

In [17]:
df['Police District'] = df['Police District'].astype(str).str.lower() 
df['Supervisor District'] = df['Supervisor District'].astype(str).str.lower() 
df['Neighborhood'] = df['Neighborhood'].astype(str).str.lower() 
df['Address'] = df['Address'].astype(str).str.lower() 
df.head()

CaseID              Opened Closed             Updated Status  \
0  11771777 2019-12-04 19:56:00    NaT 2019-12-04 20:30:10   Open   
1  11767854 2019-12-04 07:56:00    NaT 2019-12-04 08:20:36   Open   
2  11771943 2019-12-04 20:53:00    NaT 2019-12-04 21:40:05   Open   
3  11771937 2019-12-04 20:51:00    NaT 2019-12-04 21:40:04   Open   
4  11771923 2019-12-04 20:44:00    NaT 2019-12-04 21:25:05   Open   

  Status Notes Responsible Agency          Category  \
0     accepted      DPW Ops Queue  Illegal Postings   
1     accepted      DPW Ops Queue  Illegal Postings   
2     accepted      DPW Ops Queue  Illegal Postings   
3     accepted      DPW Ops Queue  Illegal Postings   
4     accepted      DPW Ops Queue  Illegal Postings   

                                 Request Type          Request Details  ...  \
0  Illegal Postings - Posted_on_Traffic_Light  Posted on Traffic Light  ...   
1       Illegal Postings - Affixed_Improperly       Affixed Improperly  ...   
2  Illegal Postings - Posted_on_Traffic_Light  Posted on Traffic Light  ...   
3  Illegal Postings - Posted_on_Traffic_Light  Posted on Traffic Light  ...   
4       Illegal Postings - Affixed_Improperly       Affixed Improperly  ...   

      Neighborhood Police District          Y           X  \
0  pacific heights        northern  37.788520 -122.430566   
1     inner sunset         taraval  37.764323 -122.466583   
2   lower nob hill         central  37.790344 -122.413654   
3   lower nob hill         central  37.790764 -122.410833   
4         nob hill         central  37.792301 -122.414105   

                        Point  Source  Media URL  Year Month Day  
0  (37.7885202, -122.4305658)     Web        NaN  2019    12   4  
1  (37.7643228, -122.4665826)     Web        NaN  2019    12   4  
2  (37.7903441, -122.4136543)     Web        NaN  2019    12   4  
3  (37.7907644, -122.4108334)     Web        NaN  2019    12   4  
4   (37.792301, -122.4141046)     Web        NaN  2019    12   4  

[5 rows x 23 columns]

In [18]:
print("Min date: ",min(df['Opened']))
print("Max date: ",max(df['Opened']))

Min date:  2008-07-03 14:56:26
Max date:  2019-12-04 20:53:00


In [20]:
df.to_csv('../../data/processed/illegalPostings.csv')